streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/AssignorConfiguration.java

```diff
@@ -108,106 +111,19 @@ public RebalanceProtocol rebalanceProtocol() {
                 case UPGRADE_FROM_21:
                 case UPGRADE_FROM_22:
                 case UPGRADE_FROM_23:
-                    // ATTENTION: The following log messages is used for verification in system test
-                    // streams/streams_cooperative_rebalance_upgrade_test.py::StreamsCooperativeRebalanceUpgradeTest.test_upgrade_to_cooperative_rebalance
-                    // If you change it, please do also change the system test accordingly and
-                    // verify whether the test passes.
-                    log.info("Eager rebalancing protocol is enabled now for upgrade from {}.x", upgradeFrom);
-                    log.warn("The eager rebalancing protocol is deprecated and will stop being supported in a future release." +
-                        " Please be prepared to remove the 'upgrade.from' config soon.");
-                    return RebalanceProtocol.EAGER;
-                case UPGRADE_FROM_24:
-                case UPGRADE_FROM_25:
-                case UPGRADE_FROM_26:
-                case UPGRADE_FROM_27:
-                case UPGRADE_FROM_28:
-                case UPGRADE_FROM_30:
-                case UPGRADE_FROM_31:
-                case UPGRADE_FROM_32:
-                case UPGRADE_FROM_33:
-                case UPGRADE_FROM_34:
-                case UPGRADE_FROM_35:
-                case UPGRADE_FROM_36:
-                case UPGRADE_FROM_37:
-                case UPGRADE_FROM_38:
-                case UPGRADE_FROM_39:
-                    // we need to add new version when new "upgrade.from" values become available
+                    log.error("The eager rebalancing protocol is no longer supported in 4.0 which means live upgrades from 2.3 or below are not possible."
+                                  + " Please see the Streams upgrade guide for the bridge releases and recommended upgrade path. Got upgrade.from='{}'", upgradeFrom);
+                    throw new IllegalStateException("Upgrading from 2.3 or below to 4.0 is not supported, got upgrade.from = " + upgradeFrom);
```

Comment:
Seems like double logging? I think we should just throw the exception and use the longer description as exception message?

And I think we should throw a `StreamsException`? We usually use standard Java exceptions only for actual bugs, but this is not a bug, but an unsupported upgrade path / user-error.

Reply:
fine with me. although I was also considering `IllegalArgumentException` or `ConfigException`, would slightly prefer one of those over a generic StreamsException, any preference?

Reply:
btw logging an error is still useful even if/when an exception is thrown since the exception message can get eaten or logged far away where it's hard to trace back to the original problem (especially in this case where we're going through consumer configuration code)